Feature/FA-10-Categories

Ammar720 opened the pull request:
This pull request introduces a new feature for managing and displaying **Categories** in the e-commerce flower app. It includes updates across the data, domain, and presentation layers to support retrieving category data from the backend and rendering it effectively in the UI.

## Changes Included:

### 🔹 Data Layer
- Added Retrofit API endpoints for fetching `categories` and `products by category`.
- Implemented a `CategoriesRemoteDataSource` to handle API calls.
- Created response models for:
  - `Category`
  - `Product`
- Developed `CategoriesRepositoryImpl` to handle data fetching and mapping logic.

### 🔹 Domain Layer
- Created use cases for:
  - Fetching all categories
  - Fetching products for a specific category
- Integrated repository contracts and abstract definitions to maintain clean separation of concerns.

### 🔹 Presentation Layer
- Added a `CategoriesCubit` to manage the business logic and states.
- Built UI screens and widgets for:
  - Displaying a list/grid of categories
  - Displaying products per selected category
- Integrated full API-UI interaction for categories feature.

---

## 📌 Pull Request Checklist

### ✅ General
- [x] This pull request is all my own work -- I have not plagiarized it.
- [x] The PR title follows the required format: `Feature/FA-10-Categories `.
- [x] The branch name follows the required format: `feature/categories`.

### 📂 Code Quality
- [x] All filenames are in `snake_case`.
- [x] All functions and variable names follow Dart naming conventions.
- [x] All new code is formatted with `dart format .`.

### 🛠 CI/CD Validation
- [x] Linting passes with `dart analyze`.
- [x] The branch name is validated against the allowed patterns.
- [x] The PR title is validated against the required format.

### 🚀 Testing
- [ ] Unit tests are included for all new features and bug fixes.
- [ ] `flutter test --coverage` has been run and the results reviewed.
- [ ] The project builds successfully without errors.

---

### 🔗 Related JIRA Ticket(s)
- [x] JIRA Issue: `FA-10`

### 📸 Screenshots (if applicable):

<table>
  <tr>
    <td><img src="https://github.com/user-attachments/assets/ce5b77ed-a016-4dca-b7ef-407dad88d24d" width="300"/></td>

<td><img src="https://github.com/user-attachments/assets/818406ed-667a-4c51-883f-1425c9c59298" width="300"/></td>   

  </tr>
</table>

<table>  
  </tr>
 <td><img src="https://github.com/user-attachments/assets/9fd28994-7080-45b9-9f3c-57ed8d16defd" width="300"/></td>

 <td><img src="https://github.com/user-attachments/assets/ef9cee35-4179-4122-a44e-77a70115fa31" width="300"/></td>
 
</table>

Commit:
feat(categories): implement categories and products data sources with Retrofit client

Commit:
feat(categories): implement repository and use cases for categories and products

Commit:
feat(categories): add CategoriesCubit and related state management for categories and products

Commit:
feat(categories): implement UI and API integration for categories feature

ali72-20 commented on lib/features/categories/data/models/categories_response/categories_response.dart:
```diff
@@ -0,0 +1,19 @@
+
+part 'categories_response.g.dart';
+
+@JsonSerializable()
```
any response model need to be nullable you don't know the data from api will be null or not

ali72-20 commented on lib/features/categories/domain/use_cases/get_categories_use_case.dart:
```diff
@@ -0,0 +1,15 @@
+import 'package:ecommerce_flower_app/features/categories/domain/repo/categories_repo.dart';
+import 'package:injectable/injectable.dart';
+
+@singleton
```
useing singlton this class will created and never removed after using it

ali72-20 replied:
use injectable insted

ali72-20 commented on lib/features/categories/presentation/view_model/categories_cubit.dart:
```diff
@@ -0,0 +1,96 @@
+import 'package:flutter_bloc/flutter_bloc.dart';
+import 'package:injectable/injectable.dart';
+
+@singleton
```
singltion can make a memory leak if you don't use this object 
insted using injectable

ali72-20 commented on lib/features/categories/presentation/view_model/categories_cubit.dart:
```diff
@@ -0,0 +1,96 @@
+          selectedTabIndex: 0,
+        ),
+      );
+  List<CategoryEntity> categories = [CategoryEntity(name: "All")];
```
make all in string insted of hard coded 
final String ALL = "All"

ali72-20 commented on lib/features/categories/presentation/view_model/categories_cubit.dart:
```diff
@@ -0,0 +1,96 @@
+  void _changeCategory(int index) {
+    emit(state.copyWith(selectedTabIndex: index));
+
+    final selectedCategoryId = index == 0 ? null : categories[index].id;
```
This assumes categories[index] will always be valid. Add a boundary check to avoid potential out-of-range errors:
if (index < 0 || index >= categories.length) return;

ali72-20 reviewed (approved):
read more about singlton and it's use can make your choice more better 
keep going

Commit:
refactor(categories): improve DI, localizations, and API response safety

- Updated localization keys (e.g. renamed 'All' → 'all')
- Marked response fields as nullable to handle API inconsistencies
- Replaced @singleton with @injectable for better memory management
- Added safe access checks for tab index in category cubit

Commit:
Merge branch 'developement' into feature/fa-10-categories

Commit:
refactor(categories): enhance loading with Skeletonizer and used CachedNetworkImage

Commit:
refactor(categories): enhance loading with Skeletonizer and used CachedNetworkImage

Commit:
Merge branch 'feature/fa-10-categories' of https://github.com/ahmedelazab1220/ecommerce_flower_app into feature/fa-10-categories

Commit:
Merge branch 'developement' into feature/fa-10-categories

Commit:
refactor(categories): simplify CategoriesScreen and enable category navigation from HomeScreen

- Refactored CategoriesScreen to accept categoryIndex via constructor
- Removed cached screen instance from MainLayoutCubit
- Switched to dynamic screen generation based on selected tab
- Implemented seamless navigation to specific category from HomeScreen
- Ensured fresh state loading and proper tab switching behavior

Commit:
Merge branch 'developement' into feature/fa-10-categories

Commit:
refactor(categories): add navigation to product details
